sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala

```diff
@@ -789,4 +789,12 @@ class DateFunctionsSuite extends QueryTest with SharedSparkSession {
         Row(Timestamp.valueOf("2015-07-24 07:00:00")),
         Row(Timestamp.valueOf("2015-07-24 22:00:00"))))
   }
+
+  test("SPARK-30668: use legacy timestamp parser in to_timestamp") {
+    withSQLConf(SQLConf.LEGACY_TIME_PARSER_ENABLED.key -> "true") {
+      val df = Seq("2020-01-27T20:06:11.847-0800").toDF("ts")
+      checkAnswer(df.select(to_timestamp(col("ts"), "yyyy-MM-dd'T'HH:mm:ss.SSSz")),
+        Row(Timestamp.valueOf("2020-01-27 20:06:11.847")))
+    }
```

Comment:
Shall we check `NULL` case together when `SQLConf.LEGACY_TIME_PARSER_ENABLED.key=false`? That will detect when we fix the regression issue in the default configuration (`SQLConf.LEGACY_TIME_PARSER_ENABLED.key=true`).

Reply:
I will add the check